hudi-spark-datasource/hudi-spark/src/main/scala/org/apache/spark/sql/hudi/command/MergeIntoHoodieTableCommand.scala

```diff
@@ -150,15 +151,35 @@ case class MergeIntoHoodieTableCommand(mergeInto: MergeIntoTable) extends Hoodie
     // Create the write parameters
     val parameters = buildMergeIntoConfig(hoodieCatalogTable)
 
+    val expectedQueryColumns = hoodieCatalogTable.tableSchemaWithoutMetaFields
+
+    val coercedQueryOutput = coerceQueryOutputColumns(StructType(expectedQueryColumns), sourceDF.logicalPlan, hoodieCatalogTable, conf)
```

Comment:
We can't do that unfortunately: `MERGE INTO` semantic does allow you to merge arbitrary input that might be drastically different from your target table, controlling what goes where w/ matched/non-matched statements (`t.id = s._other_id`, etc)

Reply:
understood Thanks @alexeykudinkin 
essentially， I realized that this issue was caused by we must use ExpressionPayload in MergeInto, but for now, write.payload.class can be overwritten by the user, so I changed the logic forced to use ExpressionPayload in MergeInto. WDYT?